### PR TITLE
Refactor Interruptible

### PR DESCRIPTION
**Pull request by anne-laure-s**
This PR is an update of this commit : https://github.com/MinaProtocol/mina/pull/12534/commits/10bea7124e9b1daa99f081c244ad9f83b02886e1

### CODE REVIEW

**Comment by anne-laure-s** [comment]
!ci-build-me

**Comment by anne-laure-s** [comment]
!ci-build-me

File: src/lib/consensus/proof_of_stake.ml [hunk 1]
```diff
@@ -830,24 +830,25 @@ module Make_str (A : Wire_types.Concrete) = struct
...
           get_delegators producer_public_key
           |> Option.value_map ~f:Hashtbl.to_alist ~default:[]
         in
         let rec go acc = function
           | [] ->
-              Interruptible.return acc
+              return acc
           | (delegator, (account : Mina_base.Account.t)) :: delegators ->
-              let%bind () = Interruptible.return () in
+              let%bind () = return () in
```
Comment (georgeee) [inline src/lib/consensus/proof_of_stake.ml:851]: I think this line can be removed

**Review by georgeee — APPROVED** [review]

File: src/lib/consensus/proof_of_stake.ml (hunk 1 above)
Comment (anne-laure-s) [reply src/lib/consensus/proof_of_stake.ml]: Removed

**Comment by anne-laure-s** [comment]
!ci-build-me

**Comment by anne-laure-s** [comment]
!ci-nightly-me

**Review by mrmr1993 — CHANGES_REQUESTED** [review]
How did this become bundled with the bitswap changes? This appears to be entirely orthogonal, and worthy of additional separate discussion.

@georgeee are there other things like this being smuggled in the bitswap work? Might be helpful to get another reviewer who wasn't also the original author as part of this PR series to make sure we're avoiding that.

**Comment by georgeee** [comment]
@mrmr1993 I disagree that it's being "smuggled". It's a PR that is entirely orthogonal and that was developed as part of Bitswap effort. It was developed because previous `Interruptible` contained unnecessarily complicated semantics which made it harder to use, way harder than necessary. Because further Bitswap work uses `Interruptible` to a great extent, I found this refactoring a good idea, back in 2022.

For precise reason why I considered this refactoring: Bit-catchup algorithm uses ivars for synchronization of deferred action results, and synchronizing current `Interruptible` with an external-used ivar would require a lot of indirection with a rather complex de-facto "mapping" to Async's scheduler. It all is replaceable by a single ivar to interrupt the computation with very clear mapping to Async scheduler.

As for "are there other things like this in the bitswap work", answer is: likely yes. Because I was developing the branch `georgee/bit-catchup` in complete isolation from `develop` of that day (wasn't a good idea, as we realized later), I also refactored some code as I saw fit. We're merging `georgeee/bit-catchup` branch as a series of self-contained PRs specifically to differentiate between edits, many of which are independent of others. There is a stack of three PRs now, but it's only for convenience, because PRs change same files and as of now we expect that all three of them are worth merging.

We can separately synchronize on the policy of reviewing changes from `georgeee/bit-catchup` branch. In case of this PR, I performed the review "from scratch". When I was doing that I didn't recall much of detail involving this PR, hence I had to perform the reviewer's work entirely from scratch, so it's not like that I wrote the code and immediately approved my own code.

-------------------------------------------------

AFAIU, your main concern is motivation to have this PR in the first place. It's perfectly valid. At the point this refactoring was performed, decision was made within then-present Networking team (consisting of Nathan and I). I'll copy-paste here messages I exchanged in Slack with Nathan on September 12, 2022, before original version of this PR was developed:

_George Agapov_

> One more thing I wanted to discuss. In my PR I'm going to use Interruptible. I've checked the code of it and usages and it seems to me unnecessarily complicated
> First, signal type is always unit . In all of the usages.
> Second, I see little point in having each monadic action having its own ivar. This is contrived and has no real effect. Because how it's used in practice is that only one action in the call chain will contain a signal-producing interruptor, others are mostly uninterruptible in that or other way

> My proposal is to streamline this a bit:
> 1. Have a module type largely the same as now, but without signal type (just unit)
> 2. Drop `lift` function
> 3. Let the interruptible be instantiated as a first-class module with one ivar created on instantiation
> 4. Check this ivar on monadic binds
> 5. Allow to interrupt this ivar any time

_Nathan Holland_

> Does each monadic action have it's own ivar? I think the interrupt ivar is passed all the way down through the monad

_George Agapov_

> Each `return`, `lift` or `uninterruptible` creates new ivar. Bind doesn't create new ivars, ofc

_Nathan Holland_

> I didn't realize uninterruptible and return were creating new ivars. We could fix that without a module functor if we wished, but I think the module functor approach better captures the usage of this library.

> You are right that most everything is uninterruptible, and I also agree it's implementation is more general than it needs to be. It's worth noting that there is a `Deferred_let_syntax` module that exists to handle this really common pattern where every bind point besides the first is a deferred wrapped in `uninterruptible`

_George Agapov_

> So in my opinion we could just use Deferred.t and have a custom Interruptible_deffered_let_syntax instantiated via first class module

> ```
> let f intr =
>   let (module Interruptible_let_syntax) = mkInterruptible intr in
>   let open Interruptible_let_syntax in
>   Deferred.return () >>= fun () -> Deferred.return ()
> 
> let g =
>   let intr = Ivar.create () in
>   don't_wait_for (f intr);
>   Ivar.fill intr ()
> ```

_Nathan Holland_

> Ok, this approach makes sense to me. It may be even better to just use a generative functor that also creates the ivar for you as well, and then you can just pull the generated module in order to send and interrupt signal (IIRC, OCaml will like this a lot more at runtime than a dynamically generated first class module).

-----------------------

So @mrmr1993, if you're convinced the design is not good, please point to its weaknesses or present other reasoning why the PR should be discarded. If concern was mainly that it occurred in bit-catchup work "by accident", I hope this message helped to explain motivation for its inclusion.

To summarize, let's consider it as an independent refactoring, that can be viewed as useful or not without relation to bit-catchup.

**Comment by mrmr1993** [comment]
> So @mrmr1993, if you're convinced the design is not good, please point to its weaknesses or present other reasoning why the PR should be discarded.

@georgeee that's not quite how this works. PRs need a good reason to be accepted, not discarded. By default, any PR's destiny is to be closed without merging.

It's hard to judge the interruptible change in the context here. It's something that needs to be considered carefully and deliberately, because it's also used to ensure safety in the transaction pool etc.
